source/sundials/arkode.cc

```diff
@@ -302,33 +294,43 @@ namespace SUNDIALS
         yy        = N_VNew_Serial(system_size);
         abs_tolls = N_VNew_Serial(system_size);
       }
-    reset(data.initial_time, data.initial_step_size, solution);
 
-    double next_time = data.initial_time;
+    DiscreteTime time(data.initial_time,
+                      data.final_time,
+                      data.initial_step_size);
+    reset(time.get_current_time(), time.get_next_step_size(), solution);
 
     if (output_step)
-      output_step(0, solution, 0);
+      output_step(time.get_current_time(), solution, time.get_step_number());
 
-    while (t < data.final_time)
+    while (time.is_at_end() == false)
       {
-        next_time += data.output_period;
+        time.set_desired_next_step_size(data.output_period);
 
-        status = SundialsARKode(arkode_mem, next_time, yy, &t, ARK_NORMAL);
+        {
+          double     t;
+          const auto status = SundialsARKode(
+            arkode_mem, time.get_next_time(), yy, &t, ARK_NORMAL);
+          (void)status;
+          AssertARKode(status);
+          time.set_desired_next_step_size(t - time.get_current_time());
+          Assert(std::abs(time.get_next_time() - t) < 1.e-11,
+                 ExcInternalError());
+        }
 
-        AssertARKode(status);
-
-        status = ARKodeGetLastStep(arkode_mem, &h);
-        AssertARKode(status);
+        time.advance_time();
 
         copy(solution, yy);
```

Comment:
Could we move this call right after the call to `SundialsARKode` (i.e., right after the error check for that call)? I think that's where it should conceptually be, not after we advance time.

Reply:
done